### DISCUSSION TITLE
Strange issue "ERROR TOCSS" - resolved?

### DISCUSSION THREAD

**Discussion by FuadEfendi** [hbstack/.github#68] [Issues and bugs]
It never happens at Netlify, but happens at my MacBook; I tried to isolate the issue.

I have project which compiles correctly on MacBook; I "shink" my bigger project to minimal healthy:

```
hugo v0.128.1+extended darwin/arm64 BuildDate=2024-07-02T06:46:41Z VendorInfo=brew


                   |  EN    
-------------------+--------
  Pages            | 38815  
  Paginator pages  |  4244  
  Non-page files   |     1  
  Static files     |     3  
  Processed images |     9  
  Aliases          | 12755  
  Cleaned          |     0  

Built in 134768 ms
```

Now, whenever I add single document to this project, such as `abc/index.md` containinng just "Hello, World" string, I get this:

```
Start building sites … 
hugo v0.128.1+extended darwin/arm64 BuildDate=2024-07-02T06:46:41Z VendorInfo=brew

ERROR TOCSS: failed to transform "/hb/modules/search/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/search@v0.2.2/assets/hb/modules/search/index.scss:2:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hugomods/search@v0.14.0/assets/search/scss/index.scss.
ERROR TOCSS: failed to transform "hb/scss/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/scss/index.tmpl.scss:4:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/twbs/bootstrap@v5.3.3+incompatible/scss/_functions.scss.

```

It seems something is hardcoded... not sure, maybe in-memory cache, happens on Mac OS only. "Magic Number" is 38815 pages; adding simple new page breaks project.

Unfortunately I cannot share this project, but maybe I can generate test project... maybe it is limitations of Hugo Embedded Web Server, so I tried to run just `hugo` command to compile project (`hugo server` will run embedded server):
```
Start building sites … 
hugo v0.128.1+extended darwin/arm64 BuildDate=2024-07-02T06:46:41Z VendorInfo=brew


                   |  EN    
-------------------+--------
  Pages            | 38816  
  Paginator pages  |  4244  
  Non-page files   |     1  
  Static files     |     3  
  Processed images |     9  
  Aliases          | 12755  
  Cleaned          |     0  

Total in 151853 ms
```

And it works!

So, finally I got it, after creating few similar tickets here (sorry!): it is limitation of Hugo Web Server.

**Comment by FuadEfendi** [comment]
I reposted it to Hugo; we can close it, I am very sure it is `hugo server` (embedded server) limitations. This explains why I have is on Mac OS (trying to run Hugo web server) and no issues at Netlify (which runs CDN instead).

https://discourse.gohugo.io/t/hugo-embedded-web-server-limitations-error-tocss/50543

**Comment by razonyang** [comment]
I'd suggest providing a minimal reproducible repo and steps, so that I can check whether if this is a theme's bug, and I don't think there will be people spend much time on this without reproducible steps.

You don't need to share your content, just remove all your private config and content, and create some dummy, and share it (make sure the issue is reproducible).

**Comment by FuadEfendi** [comment]
you used tool to generate pages; it is reproducible, but it is issue from Hugo, not from theme; I am sure it is with Hugo, just need 40,000 pages to reproduce.
Note: issue happens with `hugo server` command only; it doesn't happen when we just compile site using `hugo` command and deploy to CDN. Therefore, it is bug in Hugo, not related to theme. I posted it here just to document my finding.

**Comment by FuadEfendi** [comment]
Here it is, large site, PoC (Proof-of-Concept), I use same settings as for my other sites, generated 40,000 pages using `lorem-ipsum-generator -n 40000`

I disabled "tags" and "authors" taxonomy, but build is still slow, and the same issue, `hugo server` fails after about 10-15 minutes. **`hugo` won't fail.**

https://github.com/tokenizer-inc-canada/large-site-test

```
 large-site-test git:(main) ✗ hugo server           
go: added github.com/hugomods/masonry-js v0.1.0
go: added github.com/desandro/masonry v4.2.2+incompatible
hugo: collected modules in 631 ms
Watching for changes in /Users/fefendi/java/web/large-site-test/{archetypes,assets,content,layouts,package.json,static}
Watching for config changes in /Users/fefendi/java/web/large-site-test/config/_default, /Users/fefendi/java/web/large-site-test/config/development, /Users/fefendi/java/web/large-site-test/go.mod
Start building sites … 
hugo v0.128.1+extended darwin/arm64 BuildDate=2024-07-02T06:46:41Z VendorInfo=brew

WARN  [hugopress] [hb-custom] hook does't exist: layouts/partials/hugopress/modules/hb-custom/hooks/hb-blog-sidebar.html
ERROR TOCSS: failed to transform "/hb/modules/search/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/search@v0.2.2/assets/hb/modules/search/index.scss:2:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hugomods/search@v0.14.0/assets/search/scss/index.scss.
ERROR TOCSS: failed to transform "hb/scss/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/scss/index.tmpl.scss:4:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/twbs/bootstrap@v5.3.3+incompatible/scss/_functions.scss.
Built in 1073735 ms
Error: error building site: TOCSS: failed to transform "hb/scss/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/scss/index.tmpl.scss:4:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/twbs/bootstrap@v5.3.3+incompatible/scss/_functions.scss.
```

**Reply by FuadEfendi** [reply]
But delete randomly about 5000 pages and it won't fail. "Magic number" `38815`; maybe on Windows it will be different number.

**Comment by razonyang** [comment]
Unable to reproduce on Linux and Windows, I don't have macOS machine to test, and couldn't to reproduce on GitHub Actions since it has storage limit (14GB if I remember correctly).

![image](https://github.com/hbstack/.github/assets/17720932/bf4b334d-c287-4888-9da7-f9171a0e78eb)

![屏幕截图 2024-07-04 133117](https://github.com/hbstack/.github/assets/17720932/7fbd33ee-eaa7-449e-b591-4400068f4561)

**Reply by FuadEfendi** [reply]
I generated 40,000 pages for test project; no need to add `abc/index.md` - I did it for real-life project, but for https://github.com/tokenizer-inc-canada/large-site-test no need, it just fails on my MacBook; if I remove half of pages (any half) it doesn't fail.

**Reply by razonyang** [reply]
I just clone your repo and build, since the pages including taxonomies pages and other pages, each post generated by `lorem-ipsum-generator` may have up to 8 taxonomies (5 tags, 2 categories, 1 series) (can be duplicated) by default, so the maximum pages is about (8+1)*40k, and will be less than that if there are a lot of duplicated taxonomies.
You can try to reproduce it on other macOS machines if possible.

**Comment by FuadEfendi** [comment]
Thanks for testing; so, it seems Mac OS distribution of Hugo is broken / or, maybe some Mac internal bug

I have another error with just `hugo` command on the same project; if I delete random files issue disappears. I

**Reply by razonyang** [reply]
> it seems Mac OS distribution of Hugo is broken / or, maybe some Mac internal bug

Not sure, there are many factors may cause this, there may be environment issue, such as some permissions, disk, memory or something else.
If you have another macOS machine, please try to reproduce on it to see if this is reproducible on other macOS machines.

Btw, is your command `hugo server` without any flags?

**Comment by FuadEfendi** [comment]
Yes, I forgot, I have "smaller" MacBook Air, I can try on it too

Yes, I tried with flags `hugo server -M`, without flags `hugo server`, and it gives above mentioned error.

When I try `hugo` it gives different error. 

When I remove (random) 10000 files it doesn't give any error. 

I also tried to use "segments" but even when I try to build 1% of pages it gives error.

**Reply by FuadEfendi** [reply]
Of course could be hardware issue on my Mac. I had some issues with some files which were fixed only after I copy-pasted text into another file and deleted original... could be some non-UTF-8 characters in file for example. I'll try later on another Mac.

**Comment by FuadEfendi** [comment]
I finally run build on **different** machine, Apple Air M1 with 16Gb RAM, it took almost two hours, and it failed:

```
fefendi@Fuads-MacBook-Air large-site-test % hugo server
hugo: downloading modules …
hugo: collected modules in 54891 ms
Watching for changes in /Users/fefendi/java/web/large-site-test/{archetypes,assets,content,layouts,package.json,static}
Watching for config changes in /Users/fefendi/java/web/large-site-test/config/_default, /Users/fefendi/java/web/large-site-test/config/development, /Users/fefendi/java/web/large-site-test/go.mod
Start building sites … 
hugo v0.128.2+extended darwin/amd64 BuildDate=2024-07-04T08:13:25Z VendorInfo=brew

WARN  [hugopress] [hb-custom] hook does't exist: layouts/partials/hugopress/modules/hb-custom/hooks/hb-blog-sidebar.html
ERROR [en] REF_NOT_FOUND: Ref "/_": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:8": page not found
ERROR [en] REF_NOT_FOUND: Ref "/a": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:34": page not found
ERROR [en] REF_NOT_FOUND: Ref "/b": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:60": page not found
ERROR [en] REF_NOT_FOUND: Ref "/c": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:86": page not found
ERROR [en] REF_NOT_FOUND: Ref "/d": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:112": page not found
ERROR [en] REF_NOT_FOUND: Ref "/e": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:138": page not found
ERROR [en] REF_NOT_FOUND: Ref "/f": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:164": page not found
ERROR [en] REF_NOT_FOUND: Ref "/g": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:190": page not found
ERROR [en] REF_NOT_FOUND: Ref "/h": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:216": page not found
ERROR [en] REF_NOT_FOUND: Ref "/i": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:242": page not found
ERROR [en] REF_NOT_FOUND: Ref "/j": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:268": page not found
ERROR [en] REF_NOT_FOUND: Ref "/k": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:294": page not found
ERROR [en] REF_NOT_FOUND: Ref "/l": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:320": page not found
ERROR [en] REF_NOT_FOUND: Ref "/m": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:346": page not found
ERROR [en] REF_NOT_FOUND: Ref "/n": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:372": page not found
ERROR [en] REF_NOT_FOUND: Ref "/o": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:398": page not found
ERROR [en] REF_NOT_FOUND: Ref "/p": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:424": page not found
ERROR [en] REF_NOT_FOUND: Ref "/q": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:450": page not found
ERROR [en] REF_NOT_FOUND: Ref "/r": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:476": page not found
ERROR [en] REF_NOT_FOUND: Ref "/s": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:502": page not found
ERROR [en] REF_NOT_FOUND: Ref "/t": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:528": page not found
ERROR [en] REF_NOT_FOUND: Ref "/u": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:554": page not found
ERROR [en] REF_NOT_FOUND: Ref "/v": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:580": page not found
ERROR [en] REF_NOT_FOUND: Ref "/q": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:606": page not found
ERROR [en] REF_NOT_FOUND: Ref "/x": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:632": page not found
ERROR [en] REF_NOT_FOUND: Ref "/y": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:658": page not found
ERROR [en] REF_NOT_FOUND: Ref "/z": "/Users/fefendi/java/web/large-site-test/content/dictionary/_index.md:17:684": page not found
ERROR TOCSS: failed to transform "/hb/modules/search/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/search@v0.2.2/assets/hb/modules/search/index.scss:2:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hugomods/search@v0.14.0/assets/search/scss/index.scss.
ERROR TOCSS: failed to transform "hb/scss/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/scss/index.tmpl.scss:4:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/twbs/bootstrap@v5.3.3+incompatible/scss/_functions.scss.
Built in 6446892 ms
Error: error building site: TOCSS: failed to transform "hb/scss/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/scss/index.tmpl.scss:4:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/twbs/bootstrap@v5.3.3+incompatible/scss/_functions.scss.
fefendi@Fuads-MacBook-Air large-site-test %  

```

**Comment by FuadEfendi** [comment]
Test project has taxonomy enabled for "category" only; if I disable all taxonomies, `hugo` command will fail, but `hugo server` doesn't fail; if I add more pages, it will fail.

Probably MacOS internals, some bug, and we shouldn't wwaste time on it; documented it here so Google will find it:

```
 large-site-test git:(main) ✗ hugo
Start building sites … 
hugo v0.128.1+extended darwin/arm64 BuildDate=2024-07-02T06:46:41Z VendorInfo=brew

WARN  [hugopress] [hb-custom] hook does't exist: layouts/partials/hugopress/modules/hb-custom/hooks/hb-blog-sidebar.html
Total in 132021 ms
Error: error building site: POSTCSS: failed to transform "css/hb.css" (text/css): Error: Cannot find module '/Users/fefendi/.npm-global/lib/node_modules/@fullhuman/postcss-purgecss'
Require stack:
- /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/postcss.config.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1212:15)
    at Module._load (node:internal/modules/cjs/loader:1038:27)
    at wrapModuleLoad (node:internal/modules/cjs/loader:212:19)
    at Module.require (node:internal/modules/cjs/loader:1297:12)
    at require (node:internal/modules/helpers:123:16)
    at global.requireX (/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/postcss.config.js:10:16)
    at Object.<anonymous> (/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/purgecss.config.js:38:18)
    at Module._compile (node:internal/modules/cjs/loader:1460:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1544:10)
    at Module.load (node:internal/modules/cjs/loader:1275:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/postcss.config.js'
  ]
}

```

I can also suspect SCSS compiler issue, it is different for Windows, but I am not specialist.

**Reply by razonyang** [reply]
> Cannot find module '/Users/fefendi/.npm-global/lib/node_modules/@fullhuman/postcss-purgecss'

I think you forgot to install dependencies via `npm install`.

**Reply by FuadEfendi** [reply]
no, module is there; it is just the same issue, with 30,000 files, `hugo server` doesn't fail, but `hugo` fails; and it doesn't fail at Netlify.

I had in past issues with SAAS compiler on Windows, proper Hugo & NPM required installation of C++ DLLs etc.; in corporate environment, it was very tricky. I think it is buggy somewhere in MacOS, SAAS or something similar, too many dependencies. All errors with "large" site mention some CSS. I disabled all taxonomies, and it fails in two minutes.

**Reply by razonyang** [reply]
> I disabled all taxonomies, and it fails in two minutes.

You can try to use local node modules instead of global node modules.

```sh
cd /path/to/project
npm install @hbstack/node-packages
```

**Reply by FuadEfendi** [reply]
taking my words back, I forgot to run "npm install" ;(

Now, with 40,000 pages and taxonomy completely disable, `hugo` command builds successfully in 202 seconds, but `hugo server` fails with the same error.
```
➜  large-site-test git:(main) ✗ hugo       
Start building sites … 
hugo v0.128.1+extended darwin/arm64 BuildDate=2024-07-02T06:46:41Z VendorInfo=brew

WARN  [hugopress] [hb-custom] hook does't exist: layouts/partials/hugopress/modules/hb-custom/hooks/hb-blog-sidebar.html

                   |  EN    
-------------------+--------
  Pages            | 40144  
  Paginator pages  |   360  
  Non-page files   |     0  
  Static files     |     3  
  Processed images |     9  
  Aliases          |    69  
  Cleaned          |     0  

Total in 202536 ms


```

```
➜  large-site-test git:(main) hugo server
Watching for changes in /Users/fefendi/java/web/large-site-test/{archetypes,assets,content,layouts,package.json,static}
Watching for config changes in /Users/fefendi/java/web/large-site-test/config/_default, /Users/fefendi/java/web/large-site-test/config/development, /Users/fefendi/java/web/large-site-test/go.mod
Start building sites … 
hugo v0.128.1+extended darwin/arm64 BuildDate=2024-07-02T06:46:41Z VendorInfo=brew

WARN  [hugopress] [hb-custom] hook does't exist: layouts/partials/hugopress/modules/hb-custom/hooks/hb-blog-sidebar.html
ERROR TOCSS: failed to transform "/hb/modules/search/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/search@v0.2.2/assets/hb/modules/search/index.scss:2:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hugomods/search@v0.14.0/assets/search/scss/index.scss.
ERROR TOCSS: failed to transform "hb/scss/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/scss/index.tmpl.scss:4:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/twbs/bootstrap@v5.3.3+incompatible/scss/_functions.scss.
Built in 185739 ms
Error: error building site: TOCSS: failed to transform "hb/scss/index.scss" (text/x-scss): "/Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hbstack/hb@v0.12.3/assets/hb/scss/index.tmpl.scss:4:1": File to import not found or unreadable: /Users/fefendi/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/twbs/bootstrap@v5.3.3+incompatible/scss/_functions.scss.
➜  large-site-test git:(main) 
```

anyway, I noticed this few weeks ago, now I know that it is not malformed file but some internal third-party bug.

**Reply by FuadEfendi** [reply]
Now I realize why my 200,000+ pages site doesn't fail, with 'category' taxonomy enabled: I use local build `hugo --gc -b https://example.com` and then deploy it to Netlify by `netlify deploy --prod`.

`hugo server` fails, but we can just forgive it, it is just basic HTTP server for local tests. What I cannot forgive is that "segments" feature doesn't work, I cannot test locally 1% of site, it fails with same error message. Buils is 100x faster with 1% segment, but it still fails with same error.

**Reply by razonyang** [reply]
Hmm, I couldn't provide further help, since I couldn't reproduce on __Linux__ and __Windows__, besides I don't have a macOS machine to test, if you can provide 1) a minimal repo or 2) your segment setting (I'm not sure, I haven't use this feature yet), I can try to reproduce on GitHub Actions which is able to build on macOS runner.

Please also confirm the issue is reproducible on 1) minimal repo or 2) segmnent setting.

> GitHub Actions (standard for public repos) runner's storage limit is **14GB**, that is why I request a reproducible and minimal repo.

**Comment by razonyang** [comment]
Another idea: __if it's a platform issue__, I'd suggest giving Docker a try, such as https://docker.hugomods.com/.

**Comment by razonyang** [comment]
Now HB uses Dart Sass as the default Sass transpiler, see https://hbstack.dev/news/2024/08/migrating-to-dart-sass/.

Not sure if this help for your site, please let me know if you have time.

**Comment by FuadEfendi** [comment]
tried, precise error now...

```
Start building sites … 
hugo v0.131.0+extended darwin/arm64 BuildDate=2024-08-02T09:03:48Z VendorInfo=brew

Built in 1 ms
Error: error building site: failed to acquire a build lock: open /Users/fefendi/java/web/example.com/.hugo_build.lock: too many open files

➜  example.com git:(main) ✗ ulimit -n
1048575
```

**Reply by razonyang** [reply]
Please follow steps to check if there is template using libsass.

1. `hugo mod vendor` to cache modules into `_vendor` folder.
2. global search on your site with keyword `toCSS` (case-sensitive).
3. and check if the `toCSS` options contains `transpiler`, otherwise the default transpiler was `libsass`, not `dartsass`.
4. make sure there isn't `sass_transpiler` settings in your param configuration.

You can either update https://github.com/tokenizer-inc-canada/large-site-test, and try to rebuild to make sure it doesn't work (since I couldn't reproduce it), then push the changes, I'll check if there is something using libsass later.

**Comment by FuadEfendi** [comment]
Thank you Razon,

My update, step 0:

```
large-site-test git:(main) ✗ git reset --hard
HEAD is now at 23006e2a4 commented out most of modules to isolate test
➜  large-site-test git:(main) hugo mod get -u ./...
Update module in /Users/fefendi/java/web/large-site-test
go: upgraded github.com/hugomods/font-awesome v6.5.2+incompatible => v6.6.0+incompatible
go: upgraded github.com/hugomods/icons/vendors/font-awesome v0.6.7 => v0.6.8
go: upgraded github.com/hbstack/blog v0.37.1 => v0.37.3
go: upgraded github.com/hbstack/hb v0.12.3 => v0.12.4
go: upgraded github.com/hbstack/header v0.14.6 => v0.15.0
go: upgraded github.com/hugomods/icons/vendors/simple-icons v1.0.45 => v1.0.50
go: upgraded github.com/hugomods/simple-icons v13.1.0+incompatible => v13.6.0+incompatible
go: upgraded github.com/hbstack/hb v0.12.4 => v0.13.0
Update module in /Users/fefendi/java/web/large-site-test/node_modules/@hbstack/node-packages
Update module in /Users/fefendi/java/web/large-site-test/node_modules/netlify-cli/functions-templates/go/hello-world
➜  large-site-test git:(main) ✗ npm install          

changed 1 package, and audited 1464 packages in 6s

248 packages are looking for funding
  run `npm fund` for details

5 vulnerabilities (1 moderate, 4 high)

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
➜  large-site-test git:(main) ✗ 
```

Step 2: `hugo mod vendor`

Step 3: search for `toCSS`
```
  {{- $options := dict
    "transpiler" (default "dartsass" $params.hb.sass_transpiler)
    "targetPath" (printf "css/%s%s.css" $bundle $suffix)
    "enableSourceMap" hugo.IsProduction
    "vars" site.Params.hb.styles
  }}
 ```
  
Step 4: "make sure there isn't sass_transpiler settings in your param configuration."

Found in `_vendor` folder, comes with theme:
```
[params.hb]
css_bundle_name = "hb"
js_bundle_name = "hb"
logo = "/images/logo.png"
color = "light" # default color mode, light or dark.
sass_transpiler = "dartsass" # dartsass or libsass (deprecated).
```

**It didn't fail.**
```

                   |  EN    
-------------------+--------
  Pages            | 40085  
  Paginator pages  |   360  
  Non-page files   |     0  
  Static files     |     3  
  Processed images |     3  
  Aliases          |    41  
  Cleaned          |     0  

Built in 57038 ms
```

But my previous site was failing with **adequate** error message `too many open files` which was **good sign** too showing that something changed (different SASS implementation). It has 5x more pages so it was reasonable; but my attempt to use *huge*  `ulimit -n` didn't help>

Step 5:
Set `ulimit -n 300000` in my default `.zshrc` profile settings (MacOS has default 300, and Linux default is 1024). I assume `hugo server` was executing some commands in separate process and my settings in terminal were not inherited. Run `hugo server` for huge site again.

[Now I need to wait for at least 50 minutes... *will update in an hour*]

P.S. I initially suspected that it was "too many open files" error, improperly handled & shown as a different error "cannot open file", but unfortunately didn't dig deeper... of course 300 open files are not enough.


**[UPDATE]**

Setting `ulimit -n 300000` for MacOS default profile, and run `hugo server` in new terminal, didn't help with 280,000-pages site; same error: `Error: error building site: failed to acquire a build lock: open /Users/fefendi/java/web/example.com/.hugo_build.lock: too many open files`
Maybe sometime in a future I need to restart laptop and try again ;)

It works fine with https://github.com/tokenizer-inc-canada/large-site-test (40000 pages) site after upgrade.

**Reply by razonyang** [reply]
> It works fine with https://github.com/tokenizer-inc-canada/large-site-test (40000 pages) site after upgrade.

OK, at least, it proves that the `toCSS` issue was gone.

> too many open files... Maybe sometime in a future I need to restart laptop and try again ;)

Perhaps restart and give a much more larger number (such as 1 million) may help.

Btw, have you try the workaround from bep?

> There are some OS settings related to getting rid of the "Too many open files", but my recommended workaround is to do:
>   ```sh
>   hugo server --poll 5s
>   ```

**Comment by FuadEfendi** [comment]
Hi Razon,

Unfortunately Netlify build fails:
```
 Error: error building site: TOCSS-DART: failed to transform "hb/scss/index.scss" (text/x-scss). You need to install Dart Sass, see https://gohugo.io//functions/css/sass/#dart-sass: this feature is not available in your current Hugo version, see https://goo.gl/YMrWcn for more
```
Here is my Netlify config:
```
[build]
command = "npm ci && hugo --gc -b $URL" # change to suite your needs.
publish = "public"

[build.environment]
NODE_VERSION = "22.5.1"
HUGO_VERSION = "0.131.0"
HUGO_ENABLEGITINFO = "true"
HUGO_ENV = "production"
GO_VERSION = "1.22.5"
```

**Reply by razonyang** [reply]
Hmm, the error message is clear enough: `You need to install Dart Sass`, I'm not sure if Netlify support Dart Sass, see https://answers.netlify.com/t/hugo-builds-support-dart-sass/36531.

If Netlify doesn't support Dart Sass and doesn't failed with libsass, override the `hb.sass_transpiler` as `libsass` on `config/production/params.yaml`, to keep using libsass on production environment.

**Comment by travisbikkle** [comment]
同样的问题，hugo没有升级，在macOS和ubuntu 22.04上均失败。

ERROR TOCSS-DART: failed to transform "hb/scss/index.scss" (text/x-scss): "<stream>:1:1": connection is shut down

**Reply by razonyang** [reply]
> connection is shut down

我猜你是通过 NPM 安装的 sass，然而 JS 版的 sass 不支持 Embedded Sass protocol，建议用 Homebrew、[Snap](https://snapcraft.io/dart-sass) 或者[直接下载解压](https://github.com/sass/dart-sass/releases/tag/1.78.0) 等方式进行安装。

> The --embedded command-line flag is not available when you install Dart Sass as an [npm package](https://github.com/sass/dart-sass?tab=readme-ov-file#from-npm). No other command-line flags are supported with --embedded.
> https://github.com/sass/dart-sass?tab=readme-ov-file#embedded-dart-sass

**Reply by travisbikkle** [reply]
不是的，我是通过wget下载的https://github.com/sass/dart-sass/releases/download/1.78.0/dart-sass-1.78.0-linux-x64.tar.gz包解压到/usr/bin目录下，包括它里面的dart，并且权限都是有的。

而且我是直接clone一个新的git@github.com:hbstack/theme.git项目，跑不起来。

<img width="1275" alt="image" src="https://github.com/user-attachments/assets/047ac4ad-48af-43f7-8c4e-0c4abef65e28">

<img width="356" alt="image" src="https://github.com/user-attachments/assets/e6bed3f4-23ee-42e1-966d-200c27066acb">

我只是把 hb 从0.9 升级到了0.15，现在还原回去本地也能打包了，github action也能打包了。我猜是这些版本里面引入了一些新的scss导致需要dart scss，而这个dart scss比较蠢

**Reply by razonyang** [reply]
错误是说当前用户对于 sass 仅有执行权限，但没有读权限。不过第二张图片我看不到权限详情（没截全），另外升级后你可以[继续使用 libsass](https://zh-hans.hbstack.dev/news/2024/08/%E8%BF%81%E7%A7%BB%E8%87%B3-dart-sass/#%E5%A6%82%E4%BD%95%E7%BB%A7%E7%BB%AD%E4%BD%BF%E7%94%A8-libsass)。
